spharpy/spherical.py

```diff
@@ -1,6 +1,431 @@
 import numpy as np
 import scipy.special as special
 import spharpy.special as _special
+import logging as logger
+from pyfar import Coordinates
+from functools import lru_cache
+from spharpy.samplings.helpers import calculate_sampling_weights
+
+
+class SphHarm:
+    def __init__(self, n_max, coords, basis_type='complex', channel_convention='acn', inv_transform_type=None,
+                 normalization='n3d'):
+        r"""
+        Create a spherical harmonics class for transformation between spherical harmonics and signals.
+
+        Parameters
+        ----------
+        n_max : int,
+            Maximum spherical harmonic order
+        coords : :doc:`pf.Coordinates <pyfar:classes/pyfar.coordinates>`
+            Coordinate object with sampling points for which the basis matrix is
+            calculated
+        basis_type : str, optional
+            Type of spherical harmonic basis, either 'complex' or 'real'. The default is 'complex'.
+        channel_convention : str, optional
+            Channel ordering convention, either 'acn' or 'fuma'. The default is 'acn'.
+            (FuMa is only supported up to 3rd order)
+        inv_transform_type : str, optional
+            Inverse transform type, either 'pseudo_inverse' or 'quadrature'. The default is None.
+        normalization : str, optional
+            Normalization convention, either 'n3d', 'maxN' or 'sn3d'. The default is 'n3d'.
+            (maxN is only supported up to 3rd order)
+        """
+        self.n_max = n_max
```

Comment:
I would suggest to turn n_max and all other arguments into properties having a getter and setter. In this case `self.n_max = n_max` would call the setter, which would be the best place for performing the check that is currently done in `_validate_parameters`. Because `self.n_max` could not be used as a variable anymore, we would typically store the data in `self._n_max` indicating it as private.

Reply:
agreed